[99] Create time series grouped monthly/quarterly

Jack-0-0 opened the pull request:
Closes #99.

This PR:
* Adds function `gtr_funding_per_period` to `innovation_sweet_spots/analysis/analysis_utils.py` which replaces `gtr_funding_per_year`. This allows GtR timeseries data to be grouped by time periods other than years. Also, this can be used to achieve a more representative split as described in issue 95 by first splitting the data using `GtrWrangler.split_funding_data` into monthly data and then grouping using `gtr_funding_per_period` function by years or quarters. Closes #95.
* Adds function `gtr_get_all_timeseries_period` to `innovation_sweet_spots/analysis/analysis_utils.py` which replaces `gtr_get_all_timeseries` that can be used with monthly, quarterly and yearly data.
* Updates `innovation_sweet_spots/pipeline/pilot/timeseries_gtr.py` to use new function `gtr_get_all_timeseries_period`
* Updates functions relating to generating Crunchbase time series data in `innovation_sweet_spots/analysis/analysis_utils.py` to allow data to be grouped into months, quarters and years.
* Updates `innovation_sweet_spots/pipeline/pilot/timeseries_cb.py` to use updated functions and filter dates at the start of the processing.
* Updates `innovation_sweet_spots/tests/test_analysis_utils.py` to work with new functions
* Adds typer to `timeseries_cb.py` and `timeseries_gtr.py`  to give command line options

Commit:
Update functions to work for more time periods than year

Commit:
Update timeseries_gtr.py to use new functions

Commit:
Correct docstring to use period

Commit:
Remove unnecessary datetime conversion

Commit:
Move crunchbase filtering by year to start of process

Commit:
Update crunchbase functions to work for monthly, quarterly and yearly

Commit:
Update timeseries_cb.py to use new functions

Commit:
Add impute periods to gtr functions

Commit:
Update impute_empty_periods to work when input is empty dataframe

Commit:
Update timeseries_cb.py dir structure

Commit:
Update test_analysis_utils.py

beingkk commented:
Thanks, I will aim to review by the end of Wednesday!

beingkk commented on innovation_sweet_spots/analysis/analysis_utils.py:
```diff
@@ -75,58 +75,55 @@ def gtr_deduplicate_projects(gtr_docs: pd.DataFrame) -> pd.DataFrame:
+            'period' - time period
+            'no_of_projects' - number of new projects in a given period,
+            'amount_total' - total amount of research funding in a given period,
+            'amount_median' - median project funding in a given period
```
Hi, looks like the function does not return a column for `amount_median` anymore. Is that on purpose, or a bug?

Jack-0-0 replied:
Hi Karlis, I removed it as I wasn't using it in the forecasting project and it didn't look it was used in the ISS analysis. Do you want it to stay in?

beingkk commented on innovation_sweet_spots/pipeline/pilot/timeseries_gtr.py:
```diff
@@ -10,16 +10,18 @@
 PARAMS = import_config("iss_pilot.yaml")
-SPLIT = True
+SPLIT = False
+PERIOD = "month"
```
I guess we should keep it to `"year"` as that's what we used to report figures in the pilot project:
```suggestion
PERIOD = "year"
```

beingkk replied:
Or - alternatively - should provide a comment/instruction that this can be changed to year, month or quarter...

beingkk commented on innovation_sweet_spots/pipeline/pilot/timeseries_gtr.py:
```diff
@@ -10,16 +10,18 @@
 # Parameters (tech categories to process, time series limits)
 PARAMS = import_config("iss_pilot.yaml")
-SPLIT = True
+SPLIT = False
```
Perhaps useful to add a comment here:

```suggestion
# If SPLIT = True, then the funding amount for projects that span multiple time periods (eg, years) will be divided uniformly across those time periods
SPLIT = False
```

beingkk reviewed (approved):
Hey @Jack-0-0 , thank you very much for this comprehensive PR and for putting a lot of thought into refactoring the functions, tests, and pipeline scripts! 

It looks very good overall - I tested running it, and I've added a few comments in the code.

A more general comment: How do you see the usage of the "pipelines" after this refactoring? If I'd like to generate time series for different periods (eg, year, quarter and month) I guess I need to edit the script and re-run it? Should we instead use, eg, [command line arguments](https://www.tutorialspoint.com/python/python_command_line_arguments.htm)? (And I guess, ideally we could have a parameterised Metaflow flow here)

If you have a quick reflection about this, perhaps you can add an issue with a note for me to look into this at a later point (as you probably would like to generate the time series and move on to forecasting).

beingkk commented:
Hey, I reported the median investment deal sizes when writing up the
results. It'd be nice to keep it (unless it causes any major issues? then
perhaps makes sense to implement the median calculation as a separate
function... can leave that for me to do later)

Commit:
Add function and tests for calculating gtr median funding

Commit:
Add typer to timeseries_cb.py

Commit:
Add typer to timeseries_gtr.py and update to use median function

Commit:
Add readme for running timeseries files

Jack-0-0 commented:
> Hey @Jack-0-0 , thank you very much for this comprehensive PR and for putting a lot of thought into refactoring the functions, tests, and pipeline scripts!
> 
> It looks very good overall - I tested running it, and I've added a few comments in the code.
> 
> A more general comment: How do you see the usage of the "pipelines" after this refactoring? If I'd like to generate time series for different periods (eg, year, quarter and month) I guess I need to edit the script and re-run it? Should we instead use, eg, [command line arguments](https://www.tutorialspoint.com/python/python_command_line_arguments.htm)? (And I guess, ideally we could have a parameterised Metaflow flow here)
> 
> If you have a quick reflection about this, perhaps you can add an issue with a note for me to look into this at a later point (as you probably would like to generate the time series and move on to forecasting).

Thank you for reviewing @beingkk. Adding command line arguments for `timeseries_gtr.py` and `timeseries_cb.py` is a good idea, I've added them using typer.

Jack-0-0 commented:
> Hey, I reported the median investment deal sizes when writing up the results. It'd be nice to keep it (unless it causes any major issues? then perhaps makes sense to implement the median calculation as a separate function... can leave that for me to do later)

Ah, sorry for removing it. I've added the median calculation back in. It did need to be added as a separate function.

beingkk commented:
Thank you!!

beingkk commented:
Hey @Jack-0-0 thanks again for this! I tried running the pipeline scripts with your typer implementation and it's really cool. I think it's good to merge.
